File: files/CSEC 13.10.py
import pandas as pd

path = "files/Excel Challenge October 13th.xlsx"
input = pd.read_excel(path, usecols="B:D", skiprows=1, nrows=15)
test = pd.read_excel(path, usecols="F:G", skiprows=1, nrows=15).rename(columns=lambda x: x.replace('.1', ''))

input['Date'] = input['Date'].ffill()
input['open'] = input.groupby('Date')['Units'].transform('first')
input['Units'] = input.apply(lambda row: row['open'] + row['Units'] if pd.isna(row['Instance']) else row['Units'], axis=1)

result = input[['Date', 'Units']]

print(result.equals(test))  # True

File: files/2024-10-13/CSEC 13.10.R
library(tidyverse)
library(readxl)

path = "files/Excel Challenge October 13th.xlsx"
input = read_excel(path, range = "B2:D16")
test  = read_excel(path, range = "F2:G16")

result = input %>%
  fill(Date, .direction = "down") %>%
  group_by(Date) %>%
  mutate(open = first(Units)) %>%
  mutate(Units = ifelse(is.na(Instance) , open + Units, Units)) %>%
  select(Date, Units)

all.equal(result, test, check.attributes = FALSE)
# [1] TRUE